src/protocol/boolean/bit_decomposition.rs

```diff
@@ -0,0 +1,210 @@
+            .into_iter()
+            .map(|f_bit| q_p.clone() * f_bit)
+            .collect::<Vec<_>>();
+
+        // Step 10. [h]_B = [d + g]_B, where [h]_B = ([h]_0,...[h]_(l+1))
+        //
+        // Again, `BitwiseSum` outputs `l + 1` bits. Since [d]_B is already
+        // `l + 1` bit long, [h]_B will be `l + 2`-bit long.
+        let h_b = BitwiseSum::execute(ctx.narrow(&Step::AddDtoG), record_id, &d_b, &g_b).await?;
+
+        // Step 11. [a]_B = ([h]_0,...[h]_(l-1))
+        let a_b = h_b[0..l as usize].to_vec();
+
+        Ok(a_b)
+    }
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+enum Step {
+    GenerateRandomBits,
+    RevealAMinusB,
+    AddBtoC,
+    IsPLessThanD,
+    AddDtoG,
+}
+
+impl crate::protocol::Substep for Step {}
+
+impl AsRef<str> for Step {
+    fn as_ref(&self) -> &str {
+        match self {
+            Self::GenerateRandomBits => "generate_random_bits",
+            Self::RevealAMinusB => "reveal_a_minus_b",
+            Self::AddBtoC => "add_b_to_c",
+            Self::IsPLessThanD => "is_p_less_than_d",
+            Self::AddDtoG => "add_d_to_g",
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::BitDecomposition;
+    use crate::{
+        error::Error,
+        ff::{Field, Fp31, Fp32BitPrime, Int},
+        protocol::context::SemiHonestContext,
+        protocol::{QueryId, RecordId},
+        test_fixture::{
+            bits_to_value, join3, make_contexts, make_world, share, validate_and_reconstruct,
+            TestWorld,
+        },
+    };
+    use rand::{distributions::Standard, prelude::Distribution, rngs::mock::StepRng, Rng};
+
+    async fn bit_decomposition<F: Field>(
+        ctx: [SemiHonestContext<'_, F>; 3],
+        record_id: RecordId,
+        a: F,
+    ) -> Result<Vec<F>, Error>
+    where
+        Standard: Distribution<F>,
+    {
+        let [c0, c1, c2] = ctx;
+        let mut rand = StepRng::new(1, 1);
+
+        let s = share(a, &mut rand);
+
+        let result = join3(
+            BitDecomposition::execute(c0, record_id, &s[0]),
+            BitDecomposition::execute(c1, record_id, &s[1]),
+            BitDecomposition::execute(c2, record_id, &s[2]),
+        )
+        .await;
+
+        // bit-decomposed values must have the same bit length of the target field
+        assert_eq!(F::Integer::BITS as usize, result[0].len());
+        assert_eq!(F::Integer::BITS as usize, result[1].len());
+        assert_eq!(F::Integer::BITS as usize, result[2].len());
+
+        let bits = (0..result[0].len())
+            .map(|i| validate_and_reconstruct(&result[0][i], &result[1][i], &result[2][i]))
+            .collect::<Vec<_>>();
+
+        Ok(bits)
+    }
+
+    #[ignore]
+    #[tokio::test]
+    pub async fn fp31() -> Result<(), Error> {
+        let world: TestWorld = make_world(QueryId);
+        let ctx = make_contexts::<Fp31>(&world);
+        let [c0, c1, c2] = ctx;
+        let mut rng = rand::thread_rng();
+
+        for i in 0..10 {
```

Comment:
Instead of randomly generating 10 values, perhaps it would be better to just test a few specific edge cases like: 0, 30, 15, 16, 1. That could be faster than looping 10 times and would yield good unit test coverage.